### DISCUSSION TITLE
task_queued_timeout prevents task scheduling

### DISCUSSION THREAD

**Discussion by HajoRed** [apache/airflow#37482] [Q&A] [answered]
Hey there,

we are currently evaluating switching to airflow. In general we have already a simple setup but have problems with our use case:

* we want to use the Celery Executor as available resources are fixed
* we have a small number of worker nodes due to resource usage in each task
* each task could probably run for multiple hours even days
* triggering of tasks is always based on user requests
* currently we planned that tasks/DAGs are triggered manually with parameters which can cause multiple hundred of tasks to be scheduled at once (either in one DAG with parameters or multiple DAGs)

What we can already tell is that tasks that are triggered but not on a worker for 10 minutes are killed because of the task_queued_timeout value and our initial thought was to set this value to + infinity which is likely not a good idea.

Now some questions that did come up:

1. Is this a workflow Airflow supports/is well suited for? (possibly thousands of manually scheduled tasks, which have to stay in the queue for weeks before they could possibly be run on a worker)
2. Is the above a way we can get things to work in Airflow?
3. Where to look for further information regarding what to use setup and configuration wise for this workflow as we already did run into the problem that tasks are cancelled due to reaching the task_queued_timeout value?

**Comment by andresitorusz** [comment]
Switching to Airflow sounds like a solid move, but your use case does pose some challenges. Here's my take on your questions:

1. Airflow can handle thousands of tasks, even if they're scheduled manually and need to wait in the queue for a long time. However, the **`CeleryExecutor` might not be the best fit** for tasks that could run for days or weeks due to its default timeout settings.

2. Yes, you can make it work with Airflow, but you might need to tweak some settings. Instead of setting the `task_queued_timeout` to infinity (which isn't advisable), you could consider adjusting it to a longer duration that fits your expected task runtime. Also, you might want to explore other executors like the `KubernetesExecutor` or `CeleryKubernetesExecutor`, which might offer more flexibility for long-running tasks.

3. For further information and guidance, I'd recommend diving into the Airflow documentation, especially the section on Executors and Queue Management. Additionally, community forums or Slack channels related to Airflow could be valuable resources where you can discuss your specific use case and get advice from experienced users.

Hope this helps you navigate your Airflow setup! Let me know if you need more details on anything.

**Comment by HajoRed** [comment]
Thanks for the hints.

In addition to your suggestions, we now try to use pools for our setup. So we basically set the amount of pool slots to the available worker count and not touch the `task_queued_timeout` at all. With this we might lose some flexibility, but tasks that can not be executed right away stay then in the `scheduled` state instead of the `queued` state and don't run into the same timeout issue.

**Reply by potiuk** [reply, marked as answer]
Yep. That's precisely the case `pools` are designed for.

**Reply by potiuk** [reply]
pools I mean :)
